processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java

```diff
@@ -52,22 +55,61 @@ public class ParserUtils
     }
   }
 
-  public static Function<String, Object> getMultiValueFunction(
+  /**
+   * @return a function that processes a given string input by splitting it into multiple values
+   * using the {@code listSplitter} if thge {@code list delimiter} is present in the input. If {@code shouldParseNumbers}
+   * is enabled, the function will also try to parse any numeric values present in the input -- integers as {@code Long}
+   * and floating-point numbers as {@code Double}.
+   */
+  public static Function<String, Object> getMultiValueAndParseNumbersFunction(
       final String listDelimiter,
-      final Splitter listSplitter
+      final Splitter listSplitter,
+      final boolean shouldParseNumbers
   )
   {
     return (input) -> {
-      if (input != null && input.contains(listDelimiter)) {
+      if (input == null) {
+        return NullHandling.emptyToNullIfNeeded(input);
+      }
+
+      if (input.contains(listDelimiter)) {
         return StreamSupport.stream(listSplitter.split(input).spliterator(), false)
-                            .map(NullHandling::emptyToNullIfNeeded)
-                            .collect(Collectors.toList());
+            .map(NullHandling::emptyToNullIfNeeded)
+            .map(value -> shouldParseNumbers ? ParserUtils.tryParseStringAsNumber(value) : value)
+            .collect(Collectors.toList());
       } else {
-        return NullHandling.emptyToNullIfNeeded(input);
+        return shouldParseNumbers ?
+            tryParseStringAsNumber(input) :
+            NullHandling.emptyToNullIfNeeded(input);
+
       }
     };
   }
 
+  /**
+   * Attempts to parse the input string into a numeric value, if applicable. If the input is a number, the method first
+   * tries to parse the input number as a {@code Long}. If parsing as a {@code Long} fails, it then attempts to parse
+   * the input number as a {@code Double}. For all other scenarios, the input is returned as-is as a {@code String} type.
+   */
+  @Nullable
+  private static Object tryParseStringAsNumber(@Nullable final String input)
+  {
+    if (!NumberUtils.isNumber(input)) {
```

Comment:
i wonder if this is worth looping over the string an extra time before we do try parse attempts, or if we should just start with trying to parse it as a long. I guess having this function call saves the double tryParse which uses a regex pattern.

Reply:
Yeah, I considered something like that. However, it adds an additional overhead of regex that you note for string inputs, so I kept the current approach, which is optimized for strings